design/secrets_provider_milestone1_design.md

```diff
@@ -0,0 +1,385 @@
+| 2    | Get K8s secrets name by configuration in Secrets Provider    | Define K8S_SECRETS in `cyberark-secrets-provider` manifest<br />` name: K8S_SECRET` | - No code changes<br /><br />- Backwards compatible          | - Repetitive work<br /><br />- Not scalable |
+
+To know which K8s Secrets the apps in the namespace need we chose the first solution, Read all K8s Secrets that are marked with labels. It would be the most K8s native and avoids the repetitive work of having to list all the K8s Secrets each app needs.
+
+It is important to note however, that we are responsible for providing a solution for updating the K8s secrets and *not* how those Pods will get those secrets.
+
+##### **Customer Experience**
+
+K8s offers the option of adding labels on K8s secrets. For example, a customer would add a label on a K8s Secret like so:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: db-secret
+  labels:
+    configurable-key: env-prod # label name can be configurable
+  type: Opaque
+stringData:
+  conjur-map: |-
+    username: secrets/my_username
+```
+
+These labels do not have to be Conjur-specific and the customer would decide which label they would like to put in the K8s secrets. They would define which secrets to iterate over by K8S_SECRET_LABEL env variable in their Secrets Provider manifest. The Secrets Provider will perform a search for secrets with that label key against the [K8s](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#secret-v1-core)/[OC](https://docs.openshift.com/container-platform/4.4/rest_api/index.html#secret-v1-core) API and search for the "conjur_map" on the results.
+
+Note that just because the customer added a label for us to filter over doesn't mean that all secret entries in the K8s Secret are stored in Conjur but it gives us a way to perform a more focused search for "conjur-map" in the K8s Secrets.
+
+Secrets Provider Manifest with the K8S_SECRET_LABEL environment variable:
+
+```yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: secrets-provider-deployment
+spec:
+  replicas: 1
+  template:
+    spec:
+      serviceAccountName: sa-1
+      containers:
+      - image: secrets-provider:latest
+        name: cyberark-secrets-provider
+   ...
+      env:
+        - name: K8S_SECRET_LABELS
+          value: 'configurable-key:env-prod,configurable-key:env-test,'
+```
+
+These labels will have an OR inclusive relationship not an AND relationship. A customer can also write '*' to imply all K8s secrets in the namespace. If K8S_SECRET_LABELS is not supplied, by default, the Secrets Provider will fetch all K8s Secrets in the namespace.
+
+To read the labels set on K8s Secrets, the Service Account used for the Secrets Provider will need "list" privileges on K8s Secrets resources.
```

Comment:
another UX decrease

Reply:
IMO we are saving on UX because a customer doesn't have to list all the k8s secrets that the secrets provider has to fetch and rather can utilize a label. They will still be able to use K8S_SECRETS if they deem appropriate for their usecase and therefore don't have to add "list" permissions. We are just giving them the option

Reply:
that is correct if there is already a label in use. also - the current labels may be irrelevant. I may have 2 k8s secrets with the label `something` that only one of them has a `conjur-map` inside it. do we expect the customer to use this label because "what the heck, it's there so let's use it"?

I would really like to research on this change and ask around customers if they prefer it because i don't think it is trivial.

Also - we reduce security by giving the service-account another privilege.